### PR TITLE
Bonus

### PR DESCRIPTION
**Pull request by zolfagharipour**
### NOTE
**Make is stupid about CLI args:** extra words after the target are Make’s goals, not the script’s. For `train` with flags, run `.venv/bin/python src/logreg_train.py …` instead of `make train …`. Later we switch to use ``uv`` and wont have this problem.

----

### **Optimizers:**
New flags (optimizer, learning rate, epochs, batch size for MBGD, plot loss). Full list: `.venv/bin/python src/logreg_train.py -h`.

- **+2 BONUS — SGD / MBGD:** `--optimizer sgd` (SGD) or `--optimizer mbgd` with optional `--batch-size` (only for minibatch GD).
- **+1 BONUS — Training loss plot:** `-pl` / `--plot-loss` saves loss curves to `model/training_loss.png`.

**Example usage**
```
.venv/bin/python src/logreg_train.py dataset_train.csv -pl -e 400 -o mbgd --batch-size 10 -lr 0.0001
```

----

### **Confusion Matrix:** 
- **(+1 BONUS):** Receives **path-to-truth** and plots the confusion matrix (predictions vs. truth).

_if you really want to see the effect you should train it with **Really** low epoch and learning-rate because the data seems to be super optimize so learning is insanely fast.
Example:_
```
.venv/bin/python src/logreg_train.py dataset_train.csv -e 10 -lr 0.0001
make predict
make confusion dataset_truth.csv
```

### CODE REVIEW

**Review by ChimPansky — APPROVED** [review]
LGTM!
i played around with different epoch-counts and learning rates and could see how the accuracy went down or up.

the only case which i didnt quite understand was:
python3 src/logreg_train.py dataset_train.csv -e 1 -o sgd -lr 0.0001
with sgd even after 1 epoch we already reach 99% accuracy. is SGD so much better than the others?

Image

**Comment by zolfagharipour** [comment]
So SGD and MBGD are different than normal GD in that they have an inner loop. That means when you run SGD it loops as many times as you have data (once per data point).
In real world this should not happen with ``epoch = 1`` but apparently the dataset we have is made so perfect that learning is insanely fast (as I said before)
you can even create a dataset with 10 entry and it still learns up to **99%**. I would Say thats why evaluation is set on **98%**, we are kinda guarantied to reach there.

**Comment by ChimPansky** [comment]
> So SGD and MBGD are different than normal GD in that they have an inner loop. That means when you run SGD it loops as many times as you have data (once per data point).
> In real world this should not happen with ``epoch = 1`` but apparently the dataset we have is made so perfect that learning is insanely fast (as I said before)
> you can even create a dataset with 10 entry and it still learns up to **99%**. I would Say thats why evaluation is set on **98%**, we are kinda guarantied to reach there. 

I see. Thx for the explanation!
